django/core/files/storage.py

```diff
@@ -285,6 +285,15 @@ def _save(self, name, content):
 
         if self.file_permissions_mode is not None:
             os.chmod(full_path, self.file_permissions_mode)
+        else:
+            # If the file was moved, check that permissions are like those of a
+            # newly created file. We determine first the 'should-be' permissions
+            # for new files, then update the file perms if they don't match.
+            cur_umask = os.umask(0)
+            os.umask(cur_umask)
```

Comment:
Of separate concern to the discussion so far, but I'd like to point out the [linux umask manpage](http://man7.org/linux/man-pages/man2/umask.2.html) notes:

> It is impossible to use umask() to fetch a process's umask without at
> the same time changing it.  A second call to umask() would then be
> needed to restore the umask.  The nonatomicity of these two steps
> provides the potential for races in multithreaded programs.

> Since Linux 4.7, the umask of any process can be viewed via the Umask
> field of /proc/[pid]/status.  Inspecting this field in
> /proc/self/status allows a process to retrieve its umask without at
> the same time changing it.

I don't like the sound of races in multithreaded programs, a lot of users run multithreaded django servers. Also presumably that means lines 233-237 are already vulnerable to this race condition.